community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -192,8 +197,14 @@ public KernelTransactionImplementation( StatementOperationParts statementOperati
         this.accessCapability = accessCapability;
         this.statistics = new Statistics( this, cpuClock, heapAllocation );
         this.userMetaData = new HashMap<>();
+        AllStoreHolder allStoreHolder =
+                new AllStoreHolder( storageEngine, storageStatement, this, cursors, explicitIndexStore,
+                        this::assertOpen );
```

Comment:
Use `KernelStatement` to also check that reference count and not only termination reason. It already implements `AssertOpen` interface.

Reply:
Yes I did that but it has the unfortunate side effect that we need to acquire a kernel statement for all operations that does property reads since otherwise the reference count will be zero and we will think that the transaction has been closed.

Reply:
Fair enough. Can we work around that in some other way? Maybe by including a check on `closed` and `closing`?

Some context: 
The reason we need to do this is to protect agains reading mixed state dynamic property records as a result of property id reuse and pull updates from master. Test for this is `TerminationOfSlavesDuringPullUpdatesIT`. 

Knowing that assertOpen in `AllStoreHolder` is only used for the specific case of verifying transaction not being terminated on value property reads, checking for termination is probably good enough.

However, the knowledge dependency for seeing that this is "probably good enough" is far to complex and I would still prefer if we could verify that the transaction we are acting from is actually open and not just "not terminated".

Reply:
I am afraid I don't understand the `assertOpen` does the same thing as before except for checking the reference counting thing. It checks the `terminationReason` and throws otherwise.

Reply:
If you mean that we should add check for `closing` and `closed` we cannot - at least not for the former since currently tx hooks are run after `closing` is set to true. I would love to clean this up but I feel that this PR is not the place. As it is now we have the same behavior as before.

Reply:
Alright, I'll take that 👍